### DISCUSSION TITLE
uncertain what Dockerfile and docker-compose should have when choosing Postgres

### DISCUSSION THREAD

**Discussion by mpstaton** [payloadcms/payload#5006] [Feature Requests & Ideas]
# Documentation Issue

I chose the Postgres option in setup.  App won't start from Yarn Dev It has no trace of the word Postgres anywhere.  

## Additional Details
/* command line error log*/

20:12:07 ~/code/colearn/v01-payload/light-house $ yarn dev
yarn run v1.22.19
warning ../../../../package.json: No license field
$ cross-env PAYLOAD_CONFIG_PATH=src/payload/payload.config.ts nodemon
[nodemon] 2.0.22
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): *.*
[nodemon] watching extensions: ts
[nodemon] starting `ts-node src/server.ts -- -I`
[20:12:19] ERROR (payload): Error: cannot connect to Postgres. Details: connect ECONNREFUSED 127.0.0.1:5432
[nodemon] app crashed - waiting for file changes before starting...
^C

/* package.json */

{
  "name": "lightHouse",
  "description": "Website template for Payload",
  "version": "1.0.0",
  "main": "dist/server.js",
  "license": "MIT",
  "scripts": {
    "dev": "cross-env PAYLOAD_CONFIG_PATH=src/payload/payload.config.ts nodemon",
    "seed": "rm -rf media && cross-env PAYLOAD_SEED=true PAYLOAD_DROP_DATABASE=true PAYLOAD_CONFIG_PATH=src/payload/payload.config.ts ts-node src/server.ts",
    "build:payload": "cross-env PAYLOAD_CONFIG_PATH=src/payload/payload.config.ts payload build",
    "build:server": "tsc --project tsconfig.server.json",
    "build:next": "cross-env PAYLOAD_CONFIG_PATH=dist/payload/payload.config.js NEXT_BUILD=true node dist/server.js",
    "build": "cross-env NODE_ENV=production yarn build:payload && yarn build:server && yarn copyfiles && yarn build:next",
    "serve": "cross-env PAYLOAD_CONFIG_PATH=dist/payload/payload.config.js NODE_ENV=production node dist/server.js",
    "eject": "yarn remove next react react-dom @next/eslint-plugin-next && ts-node eject.ts",
    "copyfiles": "copyfiles -u 1 \"src/**/*.{html,css,scss,ttf,woff,woff2,eot,svg,jpg,png,js}\" dist/",
    "generate:types": "cross-env PAYLOAD_CONFIG_PATH=src/payload/payload.config.ts payload generate:types",
    "generate:graphQLSchema": "cross-env PAYLOAD_CONFIG_PATH=src/payload/payload.config.ts payload generate:graphQLSchema",
    "lint": "eslint src",
    "lint:fix": "eslint --fix --ext .ts,.tsx src",
    "payload": "cross-env PAYLOAD_CONFIG_PATH=src/payload/payload.config.ts payload"
  },
  "dependencies": {
    "@payloadcms/bundler-webpack": "^1.0.0",
    "@payloadcms/db-postgres": "^0.x",
    "@payloadcms/plugin-cloud": "^3.0.0",
    "@payloadcms/plugin-nested-docs": "^1.0.8",
    "@payloadcms/plugin-redirects": "^1.0.0",
    "@payloadcms/plugin-seo": "^1.0.10",
    "@payloadcms/richtext-slate": "^1.0.0",
    "cross-env": "^7.0.3",
    "dotenv": "^8.2.0",
    "escape-html": "^1.0.3",
    "express": "^4.17.1",
    "next": "13.5.2",
    "payload": "^2.0.0",
    "payload-admin-bar": "^1.0.6",
    "qs": "6.11.2",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-hook-form": "7.45.4",
    "react-router-dom": "5.3.4"
  },
  "devDependencies": {
    "@next/eslint-plugin-next": "^13.1.6",
    "@payloadcms/eslint-config": "^0.0.1",
    "@swc/core": "1.3.76",
    "@types/escape-html": "^1.0.2",
    "@types/express": "^4.17.9",
    "@types/node": "18.11.3",
    "@types/qs": "^6.9.8",
    "@types/react": "18.0.21",
    "@typescript-eslint/eslint-plugin": "^5.51.0",
    "@typescript-eslint/parser": "^5.51.0",
    "copyfiles": "^2.4.1",
    "eslint": "^8.19.0",
    "eslint-config-prettier": "^8.5.0",
    "eslint-import-resolver-alias": "^1.1.2",
    "eslint-plugin-filenames": "^1.3.2",
    "eslint-plugin-import": "2.25.4",
    "eslint-plugin-prettier": "^4.0.0",
    "eslint-plugin-react-hooks": "^4.6.0",
    "eslint-plugin-simple-import-sort": "^10.0.0",
    "nodemon": "^2.0.6",
    "prettier": "^2.7.1",
    "slate": "0.91.4",
    "ts-node": "10.9.1",
    "typescript": "^4.8.4"
  }
}

**Comment by jmikrut** [comment]
Hey @mpstaton — our current Dockerfile does not come with Postgres. You would need to edit your Dockerfile to include a Postgres installation, or install Postgres locally and use that. 

However, I'm going to convert this to a feature request so we can track it accordingly!

**Reply by mpstaton** [reply]
Thank you.  I'm not sure it's a feature request.  It's more like just tell me what goes in the Dockerfile.  
I do have Postgres installed on my local.  I used Docker like four years ago when I was doing my last dev project, and I would prefer not to go on a giant scavenger hunt just to set up my Dockerfile correctly.  Your docs manager should just have a copy past for Dockerfile options.  No feature necessary.

**Comment by mpstaton** [comment]
@jmikrut honestly that sounds like a rabbit hole.  What version of Postgres is it compatible with?  Why offer an option for Postgres without warnings that I would need to figure out my own installation of Postgres and cross my fingers that it works?  

I have Postgres installed, but I don't even know if I have to start it myself?  Seems like you guys are missing some helpful info somewhere.

**Reply by mpstaton** [reply]
@jmikrut do you have a paid version where I can literally just press a button and get started?  I've been trying to install and push several open source CMSs to several hosting service for days and can't get a single one to work.

**Reply by denolfe** [reply]
@mpstaton We have our Payload Cloud offering that is one-click, but as of now, only MongoDB is offered with plans for Postgres in the future.

> ECONNREFUSED 127.0.0.1:5432

As far as running locally, this indicates you do not have Postgres running or have configured a bad connection string. Double-check that the connection string that is in your .env is valid for your environment.

**Reply by paulpopus** [reply]
Hey @mpstaton if it's just local dev you're after, I've been using https://stackbricks.app/ on linux pretty successfully, very simple interface to setup local postgres DBs and close and spin them up as needed. So give it a go with that

**Reply by johngrantdev** [reply]
Hey @denolfe 
regarding  ECONNREFUSED 127.0.0.1:5432
The challenge I have found in building the dockerfile with a local postgres docker container is that the db uri can be inaccessible from within the container without specificly passing the host network (if postgres is running on the host) or defining a network for inter container communication between payload and postgres containers. The payload build script requires a db connection which will cause the build to fail after a long wait downloading the dependencies. my preference would be to be able to skip db connection when building so that the db can be set as an environment variable for the container when the container is configured and run in the hosting environment. Is this feasable? does payload need a db connection during build?

**Reply by denolfe** [reply]
@jayelg As of 2.8.1, initializing payload can build without the need to initialize. Can you see if this version or later achieves what you desire?

**Reply by johngrantdev** [reply]
@denolfe I'm currently on 2.11.0 but still experiencing it. Looking into it, I think the issue for me is related to my project using the website-template architecture where the nextjs build starts the server.
A solution I found was to update the next:build script
from:
`"build:next": "cross-env PAYLOAD_CONFIG_PATH=dist/payload/payload.config.js NEXT_BUILD=true node dist/server.js",`
to:
`"build:next": "cross-env PAYLOAD_CONFIG_PATH=dist/payload.config.js NEXT_BUILD=true next experimental-compile",
`
next experimental-compile is discussed [here](https://github.com/vercel/next.js/discussions/46544#discussioncomment-7115457).
you then may need to run next experimental-generate after build, I've not experimented with this too much.

**Reply by ardiansyaherwin** [reply]
@johngrantdev Hi, currently I'm facing the same issue when try to deploy to Northflank. Would you mind to share your Dockerfile as well as the scripts section in your package.json

**Reply by johngrantdev** [reply]
@ardiansyaherwin 
you could look at this
https://github.com/johngrantdev/personal-site/blob/main/package.json
There I added build:skipssg and build:next:skipssg scripts that use this. I'm not using this approach right now and was looking at this a while ago so my dockerfile doesn't reflect this approach.
My thinking is you might create an entry point script for the docker file that runs npx next experimental-generate on first run before starting the server. hopefully there is something there that might work for you.

**Reply by ardiansyaherwin** [reply]
@johngrantdev thanks for the response.
I'm not sure how to edit the Dockerfile.

is it something like this?
```
COPY . .
RUN npx next experimental-generate // adding the script before build
RUN yarn install
RUN yarn build // should I also update this command to yarn build:skipssg
```

**Reply by johngrantdev** [reply]
@ardiansyaherwin 
The idea is to delay the static site generation from happening during the docket build process and instead for SSG to happen on the hosting server with all of the environment variables right.
So you wouldn't RUN npx next experimental-generate   when building the docket image.
Yes you'd RUN yarn build:skipssg  in the docker file to skip SSG which would otherwise pick up environment variables from the development environment. 
You would then either manually run npx next experimental-generate inside the docket container when it first runs on the hosting server or you could make an entry point script in your project that does this automatically if it detects a first run.

**Reply by ardiansyaherwin** [reply]
@johngrantdev i see, ok let me do more research. Thank you so much

**Comment by ArthurSav** [comment]
As a new user that was just trying to make things run, this doesn't inspire much confidence. 

I followed the quick start guide, chose Postgres and then the app fails to start. No mention that I should setup my own DB first or what is expected.

**Reply by r1ft1** [reply]
Yeah I currently use Payload in my team and thought I'd use it again for a personal project, but I'm confused seeing the [Out of the box](https://payloadcms.com/docs/getting-started/what-is-payload) statement listing a Database, when seemingly I have to spin it up myself?

**Comment by eaglecode5** [comment]
Agreed with the sentiment above.  The docs are a bit unclear.  I create the app, jump over to the app-configuration steps, but no mention of how to set up the postgres connection or configuring with Docker.

Nothing I can't research and learn, but you see where I'm coming from.
